### PR TITLE
Layout.FromLayoutMethod to allow custom layoutrenders using Config API

### PR DESCRIPTION
**Pull request by snakefoot**
Now also with ThreadAgnostic-option. Follow up to  #3737

### CODE REVIEW

File: src/NLog/Layouts/Layout.cs [hunk 1]
```diff
@@ -142,12 +142,20 @@ public static Layout CreateFromMethod(Func<LogEventInfo, object> layoutMethod, L
...
             var name = $"{layoutMethod.Method?.DeclaringType?.ToString()}.{layoutMethod.Method?.Name}";
 #endif
-            var layoutRenderer = ((options & LayoutRenderOptions.ThreadSafe) != 0) ?
-                new LayoutRenderers.FuncThreadSafeLayoutRenderer(name, (l, c) => layoutMethod(l)) :
-                new LayoutRenderers.FuncLayoutRenderer(name, (l, c) => layoutMethod(l));
+            var layoutRenderer = CreateFunLayoutRenderer(layoutMethod, options, name);
             return new SimpleLayout(new[] { layoutRenderer }, layoutRenderer.LayoutRendererName, ConfigurationItemFactory.Default);
         }
 
+        private static LayoutRenderers.FuncLayoutRenderer CreateFunLayoutRenderer(Func<LogEventInfo, object> layoutMethod, LayoutRenderOptions options, string name)
```
Comment (304NotModified) [inline src/NLog/Layouts/Layout.cs]: Func? Fun? ;)

File: src/NLog/LayoutRenderers/FuncThreadAgnosticLayoutRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+{
+    using System;
+    using NLog.Config;
+
+    /// <summary>
+    /// A layout renderer which could have different behavior per instance by using a <see cref="Func{TResult}"/>.
+    /// </summary>
+    [ThreadSafe]
+    [ThreadAgnostic]
+    class FuncThreadAgnosticLayoutRenderer : FuncLayoutRenderer
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/FuncThreadAgnosticLayoutRenderer.cs]: Please add `internal`, thanks!

File: src/NLog/Layouts/LayoutRenderOptions.cs [hunk 3]
```diff
@@ -49,5 +49,9 @@ public enum LayoutRenderOptions
         /// Layout renderer method can handle concurrent threads
         /// </summary>
         ThreadSafe = 1,
+        /// <summary>
+        /// Layout renderer method is agnostic to current thread context
```
Comment (304NotModified) [inline src/NLog/Layouts/LayoutRenderOptions.cs]: I would be nice if we could describe this without the word "agnostic"

File: src/NLog/LayoutRenderers/FuncThreadAgnosticLayoutRenderer.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/FuncThreadAgnosticLayoutRenderer.cs]: Fixed

File: src/NLog/Layouts/LayoutRenderOptions.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Layouts/LayoutRenderOptions.cs]: Tried to improve the xml-docs

**Comment by snakefoot** [comment]
Updated wiki: https://github.com/NLog/NLog/wiki/Configure-from-code

**Pull request merged**